### PR TITLE
[image_picker] Fixes an issue where videos could be wrongly picked with .jpg extension

### PR DESCRIPTION
**Pull request by miguelpruivo**
Fixes an issue where videos could be wrongly picked with `.jpg` extension.

## Description

When picking videos on Android devices, it will return file names ended with `.jpg` extension, regardless of the file being actually a video. This requires a workaround implementation, such as opening a File descriptor on Flutter side and rename it to `.mp4` again.

The issue happens because some URI paths doesn't have an extension and it was being extracted just by splitting the file name from `.` and getting the last part, which make it not quite reliable for some scenarios. Right now, I've updated the source to retrieve the extension based on the URI content.

As a note, this doesn't happen with all Android devices, but I can see it easily replicable with Samsung phones.

## Related Issues

Fixes [#52419](https://github.com/flutter/flutter/issues/52419)

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [X] All existing and new tests are passing.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [X] I read and followed the [Flutter Style Guide].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [X] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [X] I updated CHANGELOG.md to add a description of the change.
- [X] I signed the [CLA].
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/image_picker/image_picker/pubspec.yaml
Comment (awazgyawali) [inline packages/image_picker/image_picker/pubspec.yaml:5]: I think 0.6.7+4 would make more sense.

File: packages/image_picker/image_picker/CHANGELOG.md
Comment (awazgyawali) [inline packages/image_picker/image_picker/CHANGELOG.md:1]: Same here

**Comment by miguelpruivo** [comment]
This has been approved by all reviewers. If it could be merged it would be great to prevent constant rebasing 😄

Thanks!

**Comment by agavrel** [comment]
Could this PR be approved? Many thanks in advance

**Comment by abner-hello-world** [comment]
I need this fix, and I have this problemI hope the author will update soon. Thank you

**Comment by venkatd** [comment]
I can confirm that running off the fork resolves the issue for me

```yaml
  image_picker:
    git:
      url: https://github.com/miguelpruivo/plugins
      ref: 9cac5347d9ab9d111a918a8154fb060565b1bf31
      path: packages/image_picker/image_picker
```

**Comment by abner-hello-world** [comment]
Thank you for your email, which solved my problem.
| |
张涛
|
|
abner_hello_world@163.com
|
签名由网易邮箱大师定制


On 08/6/2020 02:35，Venkat Dinavahi<notifications@github.com> wrote：

I can confirm that running off the fork resolves the issue for me

image_picker:
    git:
      url: https://github.com/miguelpruivo/pluginsref: 9cac5347d9ab9d111a918a8154fb060565b1bf31path: packages/image_picker/image_picker

**Comment by SivaramNalliboyana** [comment]
> I can confirm that running off the fork resolves the issue for me
> 
> My pubspec.yaml declaration
> 
> ```yaml
>   image_picker:
>     git:
>       url: https://github.com/miguelpruivo/plugins
>       ref: 9cac5347d9ab9d111a918a8154fb060565b1bf31
>       path: packages/image_picker/image_picker
> ```

I tried the way, but still it is giving me a .jpg

**Comment by pedromassango** [comment]
@miguelpruivo You may want to add `Fixes` before the issue id so that it automatically closes the issue when merged.

**Comment by miguelpruivo** [comment]
@pedromassango done! 🎉

**Comment by PixelToast** [comment]
Hi, @danielwalczak @awazgyawali @fdcgr

Thanks for your reviews!

As per our tree hygiene https://github.com/flutter/flutter/wiki/Tree-hygiene#how-to-review-code we avoid people without commit access to this repo approving changes during a review as this confuses authors into thinking the PR is ready to merge. (Regular reviews are much appreciated, just don't click the approve changes button)

@miguelpruivo

Very sorry for such a long wait 😢

I would love to merge this however due to the lack of unit tests and my lack of experience here are some questions:

1. Why does this happen? Does this only apply to videos not saved with a file extension?
1. How hard would it be to add tests?
1. How does this behavior differ from iOS?
1. This fix seems to replace the extension detection logic entirely rather than using it as a fallback. If the uri does have an extension will this still use that extension verbatim? I'm specifically looking for any notable changes in behavior outside of files given `.jpg`.

I trust that it works because you have done this sort of thing before but we still need to make sure this doesn't break anything on a patch release version. (love your work on file_picker btw)

**Comment by ditman** [comment]
@miguelpruivo I found another approach to do this here: https://github.com/flutter/plugins/pull/2821, do you think this is something that can be collaborated upon, so we only have to merge one PR to fix two bugs?

**Comment by pedromassango** [comment]
I agree that the fix for image extension issue is better handled here than on https://github.com/flutter/plugins/pull/2821. Maybe collaborate is the better idea to merge just one PR.

**Comment by miguelpruivo** [comment]
@PixelToast hi, thank you for the input. 

Currently I’m off on vacations until next week. However, I’ll try to help you as much as I can. 

- Apparently this happens because the uri is treated as an image and not a video;
- Unit tests are not actually the easiest for native pickers as you’ll have to fully drive the picking behavior (integration test) and handle its results;
- This only affects Android, on iOS the plugin implementation is behaving properly; 

Let me know if you need anything else.

**Comment by ened** [comment]
@PixelToast @miguelpruivo @pedromassango 

Hi! I prefer to have two PRs, one for the image resizing (GIF) and one for the correct/incorrect JPG extension handling.

@ditman Let's get *this* PR ready for merge first and I can rebase #2821 after that.

**Comment by ditman** [comment]
@ened sounds good to me!

**Comment by PixelToast** [comment]
@miguelpruivo Thanks for the update, it's good to know you are still able to help.

If full integration tests aren't feasible, just a unit test for getImageExtension should be fine.

**Comment by miguelpruivo** [comment]
@PixelToast ok no problem. I’ll try to get it done as soon as I return home on the weekend.

**Comment by nicemanis** [comment]
Tested with:
```
image_picker:
  git:
    url: https://github.com/miguelpruivo/plugins
    ref: 9cac5347d9ab9d111a918a8154fb060565b1bf31
    path: packages/image_picker/image_picker
```
Videos are selected correctly with .mp4 extension. The picked file playback also works as expected.

**Comment by danielwalczak** [comment]
Go go ! 🎉🎉

**Comment by hl9565** [comment]
thanks!
it fixed my problem.
hope merge this pr soon~

**Comment by pedromassango** [comment]
@miguelpruivo this may need a rebase from master.

**Comment by miguelpruivo** [comment]
Done.

**Review by jpetro416 — APPROVED** [review]
Correctly fixes extension issue

**Comment by pedromassango** [comment]
This seems ready for review.

Cc @cyanglaz

**Comment by frankyvij** [comment]
Can this be reviewed and merged, it has been 5 days already.

**Comment by antonvinceguinto** [comment]
> I can confirm that running off the fork resolves the issue for me
> 
> My pubspec.yaml declaration
> 
> ```yaml
>   image_picker:
>     git:
>       url: https://github.com/miguelpruivo/plugins
>       ref: 9cac5347d9ab9d111a918a8154fb060565b1bf31
>       path: packages/image_picker/image_picker
> ```

Help, this works on debug but gives me an error when building release
image

**Comment by miguelpruivo** [comment]
@frankyvij it has been 3 months, actually. 😅

**Review by cyanglaz — COMMENTED** [review]
Thanks for the PR! This is great. We will need to add tests before we can land this.

**Comment by syssam** [comment]
any update for this pr? thanks

**Comment by tegarkurniawan** [comment]
any update ?

**Review by do4Mother — APPROVED** [review]

**Review by do4Mother — CHANGES_REQUESTED** [review]
i test it using latest stable flutter version. and its works fine

**Comment by thegrxp** [comment]
Any reason why it's not merged yet?

**Comment by eugrro** [comment]
Can this be merged in? This seems to be a problem that has existed for a while

**Comment by miguelpruivo** [comment]
@pedromassango I'm sorry but currently I'm out of time to work on that. I might implement it but with no ETA.

**Comment by danielroek** [comment]
I created a PR with a working Unit Test. #3685

**Comment by cyanglaz** [comment]
closing this in favor of https://github.com/flutter/plugins/pull/3685 as that one has tests. 
We are going to try to land https://github.com/flutter/plugins/pull/3685 asap

**Pull request closed without merging**
